src/main.rs

```diff
@@ -218,45 +320,150 @@ async fn apply(
         }
     };
 
+    let all_eips = eip_api.list(&ListParams::default()).await?.items;
+    let eip = all_eips
+        .into_iter()
+        .find(|eip| &eip.spec.pod_name == pod_name)
+        .ok_or_else(|| Error::NoEipResourceWithThatPodName(pod_name.to_owned()))?;
+    let eip_name = eip.metadata.name.as_ref().ok_or(Error::MissingEipName)?;
+    let allocation_id = eip
+        .status
+        .as_ref()
+        .ok_or(Error::MissingEipStatus)?
+        .allocation_id
+        .as_ref()
+        .ok_or(Error::MissingAllocationId)?;
     let eip_description = eip::describe_address(ec2_client, allocation_id.to_owned())
         .await?
         .addresses
         .ok_or(Error::MissingAddresses)?
         .swap_remove(0);
+    let public_ip = eip_description.public_ip.ok_or(Error::MissingPublicIp)?;
     if eip_description.network_interface_id != Some(eni_id.to_owned())
         || eip_description.private_ip_address != Some(pod_ip.to_owned())
     {
         eip::associate_eip_with_pod_eni(
             ec2_client,
             allocation_id.to_owned(),
-            eni_id,
+            eni_id.to_owned(),
             pod_ip.to_owned(),
         )
         .await?;
     }
+    set_eip_status_attached(eip_api, eip_name, eni_id, pod_ip.to_owned()).await?;
     add_dns_target_annotation(
         pod_api,
         pod_name.to_owned(),
-        public_ip.to_owned(),
-        allocation_id,
+        public_ip,
+        allocation_id.to_owned(),
     )
     .await?;
     Ok(ReconcilerAction {
         requeue_after: Some(Duration::from_secs(300)),
     })
 }
 
-/// Deletes AWS Elastic IP associated with a pod being destroyed.
-async fn cleanup_pod_eip(ec2_client: &Ec2Client, pod: Pod) -> Result<ReconcilerAction, Error> {
-    info!("Cleaning up...");
-    let pod_uid = pod.metadata.uid.as_ref().ok_or(Error::MissingPodUid)?;
+async fn apply_eip(
+    ec2_client: &Ec2Client,
+    eip_api: &Api<Eip>,
+    eip: Eip,
+    cluster_name: &str,
+    default_tags: &HashMap<String, String>,
+) -> Result<ReconcilerAction, Error> {
+    let eip_uid = eip.metadata.uid.as_ref().ok_or(Error::MissingEipUid)?;
+    let eip_name = eip.metadata.name.as_ref().ok_or(Error::MissingEipName)?;
+    let pod_name = eip.spec.pod_name;
     let addresses =
-        &eip::describe_addresses_with_tag_value(ec2_client, eip::POD_UID_TAG, pod_uid.to_owned())
+        eip::describe_addresses_with_tag_value(ec2_client, eip::EIP_UID_TAG, eip_uid.to_owned())
             .await?
             .addresses
             .ok_or(Error::MissingAddresses)?;
-    for address in addresses {
-        eip::disassociate_and_release_address(ec2_client, address).await?;
+    let (allocation_id, public_ip) = match addresses.len() {
+        0 => {
+            let response = eip::allocate_address(
```

Comment:
Here's a fun question that just occurred to me (not for this PR, but it's related): what kinds of indication do we have if the EIP operator is running into resource limits (i.e., can't allocate any more EIPs)? Does it have a way to report errors such that we can see them, or is "can't spin up pods with EIPs" the only indicator we get?

Reply:
It will show the AWS SDK error in the logs, but that's the only way at the moment. We don't have any alerting on it. This is definitely a weak point.

The logs are also extremely messy, especially after this change, since everything is async. I want to move to `tracing` and push the spans up somewhere (honeycomb? sentry?), but I felt that was too much to cram into this one PR.

As for running out of EIPs in AWS, we should add some dedicated monitoring for that, so we can request an increase before running out, rather than reacting to failures. There is an open ticket for this somewhere.